File: leetcode-in-golang/algo/sort/select_sort.go
package sort

func selectSort(nums []int, lo, hi int) {
	for i := lo; i < hi; i++ {
		minIndex := i
		for j := i; j < len(nums); j++ {
			if nums[j] < nums[minIndex] {
				minIndex = j
			}
		}
		nums[i], nums[minIndex] = nums[minIndex], nums[i]
	}
}

func SelectSort(nums []int) {
	selectSort(nums, 0, len(nums))
}

File: leetcode-in-rust/src/algo/sort/select_sort.rs
pub fn select_sort<T: PartialOrd + Copy>(nums: &mut Vec<T>) {
    for i in 0..nums.len(){
        let mut min_index = i;
        for j in i..nums.len(){
            if nums[j] < nums[min_index]{
                min_index = j;
            }
        }
        nums.swap(i, min_index);
    }
}


#[cfg(test)]
mod tests {
    use super::*;
    use crate::algo::sort::helper::is_sorted;

    #[test]
    fn test_is_sorted() {
        let test_case = vec![
            vec![0, 0, 0],
            vec![1, 2, 4, 5],
            vec![3, 3],
            vec![3, 1, 3],
            vec![3, 1, 0],
        ];

        for mut nums in test_case {
            select_sort(&mut nums);
            assert_eq!(is_sorted(nums), true);
        }
    }
}
